### DISCUSSION TITLE
How best to think of protocols

### DISCUSSION THREAD

**Discussion by dadepo** [libp2p/rust-libp2p#2686] [Q&A]
While going through the source, I found the list of protocols [here](https://github.com/libp2p/rust-libp2p/tree/master/protocols)

I am a bit confused on how to think about protocols semantically in `rust-libp2p` as I see things that represents different things conceptually. For example in the directory there is `kad` and `mdns`, which I believe are peer discovery mechanism. Also there is `ping` and `identify` which I believe has nothing to do with discovery but are closer to the definition of what a protocol should be. Also there is `floodsub` and `gossipsub` which appears to be conceptually different also.

So what exactly are protocols in `rust-libp2p`? What is the best way to think about them conceptually?

**Comment by thomaseizinger** [comment]
I think the [Wikipedia definition](https://en.wikipedia.org/wiki/Communication_protocol) is pretty good here:

> A communication protocol is a system of rules that allows two or more entities of a [communications system](https://en.wikipedia.org/wiki/Communications_system) to transmit [information](https://en.wikipedia.org/wiki/Information) via any kind of variation of a [physical quantity](https://en.wikipedia.org/wiki/Physical_quantity). The protocol defines the rules, [syntax](https://en.wikipedia.org/wiki/Syntax), semantics and [synchronization](https://en.wikipedia.org/wiki/Synchronization) of [communication](https://en.wikipedia.org/wiki/Communication) and possible [error recovery methods](https://en.wikipedia.org/wiki/Error_detection_and_correction).

There are also [specifications](https://github.com/libp2p/specs) for the libp2p protocols.

In a nutshell, you can think of every communication that can be defined in an isolated fashion as a protocol. On this level, `kad`, `mdns`, `ping`, `identify`, `gossipsub` are all protocols. They may allow you to achieve different goals, like peer discovery, metadata exchange or messaging but fundamentally, they are all protocols. In fact, every communication between two pieces of software IS a protocol. Often times, it is just not as explicitly specified.

**Reply by mxinden** [reply]
The mental model I use:

`/transports` contains transport protocols, enabling one to exchange bytes between **two endpoints only**.
`/protocols` contains peer-to-peer protocols, building on top of some mechanism to exchange bytes between two endpoints, orchestrating these connections to build something bigger **beyond two peers**.

Does that help?
